File: solutions/ex1_bcast.c
#include <stdio.h>
#include <mpi.h>
#define NMAX 10
#define max(i,j) ( i > j ? i : j )
#define min(i,j) ( i > j ? j : i )

int main(int argc, char *argv[])
{

int i,ierr,proc,nproc, A[NMAX];
int n;

ierr = MPI_Init(&argc,&argv);

ierr = MPI_Comm_size(MPI_COMM_WORLD, &nproc);
ierr = MPI_Comm_rank(MPI_COMM_WORLD, &proc);
if (proc==0) {
   printf("Enter the value of n, between 1 and %i\n",NMAX);
   scanf("%i",&n);
   n=min(NMAX,max(1,n));
   for (i = 0; i < n; i++)
      A[i]=(i+1)*(i+1);
}
ierr = MPI_Bcast(&n,1,MPI_INT,0,MPI_COMM_WORLD);
ierr = MPI_Bcast(A,n,MPI_INT,0,MPI_COMM_WORLD);

printf("proc %i A = ", proc);
for (i = 0; i < n; i++)
   printf("%i ", A[i]);
printf("\n");

ierr = MPI_Finalize();

return 0;
}

File: solutions/ex1_bcast.f90
program bcast
use mpi_f08
integer, parameter :: nmax=10
integer i,ierr,n,proc,nproc, A(nmax)

call MPI_Init()
call MPI_Comm_size(MPI_COMM_WORLD, nproc)
call MPI_Comm_size(MPI_COMM_WORLD, proc)
if (proc==0) then
   print '(A,I4)', 'Enter de value of n, between 1 and',nmax
   read *, n
   n=min(nmax,max(1,n))
   do i=1,n
      a(i)=i**2
   enddo
endif
call MPI_Bcast(n,1,MPI_INTEGER,0,MPI_COMM_WORLD)
call MPI_Bcast(A,n,MPI_INTEGER,0,MPI_COMM_WORLD)
print *, 'proc ',proc,' A(1:n) = ',A(1:n)
call MPI_Finalize()
end program bcast
